constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -269,7 +276,60 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display> QuadraticSymbolicExp
             }
         } else {
             self.solve_affine(range_constraints)?
-        })
+        };
+        if let Some(disjoint_set) = self.disjoint_set(range_constraints) {
+            result.effects.extend(disjoint_set);
+        }
+        Ok(result)
+    }
+
+    fn disjoint_set(
+        &self,
+        range_constraints: &impl RangeConstraintProvider<T, V>,
+    ) -> Option<Vec<Effect<T, V>>> {
+        if self.is_quadratic() {
+            None
+        } else {
+            let (flags, num_active) = self.sum_of_flags(range_constraints)?;
+            if num_active.is_one() {
+                Some(vec![Effect::DisjointSet(DisjointSet {
+                    variables: flags.into_iter().collect(),
+                })])
+            } else if num_active.is_zero() {
+                Some(
+                    flags
+                        .into_iter()
+                        .map(|v| Effect::Assignment(v, T::zero().into()))
+                        .collect::<Vec<_>>(),
+                )
+            } else {
```

Comment:
In theory you could have a set of n flags where exactly k <= n are 1, right? I'm curious if this is used anywhere, I assume not.

Reply:
I'm not aware of that but there are some very wild constructions like [this one](https://github.com/openvm-org/openvm/blob/v1.0.0/extensions/rv32im/circuit/src/loadstore/core.rs#L113-L119):
- The flags are 0, 1, or 2
- The sum is 1 or 2

This leaves exactly 14 possible configuration, and there are expressions to select each of them.

So yeah, I think it's worth considering if we can generalize this to a degree that we could solve this as well.